aten/src/ATen/native/cudnn/Conv_v8.cpp

```diff
@@ -76,99 +139,533 @@ void filterEngineConfigs(
+  uint8_t z_alignment;
+  uint8_t b_alignment;
+  // TODO: does it make sense to have this in the key? but alpha is a graph-level param...
+  float alpha;
+};
 
-}
+template <typename T, typename KeyType>
+struct BenchmarkCache {
+std::mutex mutex;
+std::unordered_map<KeyType, cudnn_frontend::ExecutionPlan, ParamsHash<KeyType>, ParamsEqual<KeyType>> engine_cache;
 
-void raw_cudnn_convolution_forward_out(
-    const Tensor& output, const Tensor& input, const Tensor& weight,
-    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups,
-    bool benchmark, bool deterministic, bool allow_tf32)
-{
-  TORCH_CHECK(!benchmark, "not supported yet");
-  if (output.numel() == 0) {
-    return;
+// TODO: is this thread safe if cache is updated? is pointer stale?
+cudnn_frontend::ExecutionPlan* find(const KeyType& key) {
+  std::lock_guard<std::mutex> guard(mutex);
+  auto it = engine_cache.find(key);
+  if (it == engine_cache.end()) {
+    return NULL;
   }
+  // TODO: probably want ExecutionPlan copy constructor or better way to return
+  return &(it->second);
+}
 
-  cudnnHandle_t handle = getCudnnHandle();
+void emplace(const KeyType& key, T& results) {
+  std::lock_guard<std::mutex> guard(mutex);
+  engine_cache.emplace(key, std::move(results));
+}
 
-  CacheKey key;
-  setConvolutionParams(&key.params, input, weight, padding, stride, dilation, groups, deterministic, allow_tf32);
-  key.input_alignment = getAlignment(input);
-  key.output_alignment = getAlignment(output);
-  key.weight_alignment = getAlignment(weight);
-
-  auto run = [&](cudnn_frontend::ManagedOpaqueDescriptor cfg) {
-    auto plan = cudnn_frontend::ExecutionPlanBuilder()
-        .setHandle(handle)
-        .setEngineConfig(cfg)
-        .build();
-
-    auto workspace_size = plan.getWorkspaceSize();
-    auto workspace = at::empty({workspace_size}, input.options().dtype(kByte));
-    void *data_ptrs[] = {input.data_ptr(), output.data_ptr(), weight.data_ptr()};
-    // std::cout << plan.describe() << " requires workspace " << workspace_size << std::endl;
-    int64_t uids[] = {'x', 'y', 'w'};
-    auto variantPack = cudnn_frontend::VariantPackBuilder()
-        .setWorkspacePointer(workspace.data_ptr())
-        .setDataPointers(3, data_ptrs)
-        .setUids(3, uids)
-        .build();
-    AT_CUDNN_CHECK(cudnnBackendExecute(handle, plan.get_raw_desc(), variantPack.get_raw_desc()));
-  };
+};
 
-  auto search = engine_cache.find(key);
-  if (search != engine_cache.end()) {
-    run(search->second);
-    return;
-  }
+BenchmarkCache<cudnn_frontend::ExecutionPlan, CacheKey> benchmark_cache;
+BenchmarkCache<cudnn_frontend::ExecutionPlan, CacheKeyFused> benchmark_cache_fused;
+
+} // namespace
+void get_cachekey(CacheKey& key, const cudnnBackendDescriptorType_t operation, const Tensor& y, const Tensor& x, const Tensor& w, const IntArrayRef padding, const IntArrayRef stride, const IntArrayRef dilation, int64_t groups, bool deterministic, bool allow_tf32) {
```

Comment:
why is this a void function instead of returning CacheKey?

Reply:
We wanted to be extra-safe here, basically it isn't clear if something like `CacheKey mykey = get_cachekey(...)` would copy "unused" padding bytes in addition to bytes in fields. This has bitten us before as we use `ParamsHash` https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/utils/ParamsHash.h for the `unordered_map` which assumes two `CacheKeys` to be identical only if they are bit-perfect.

Reply:
Then you should define constructor of CacheKey that sets extra bits to 0 and that you can directly call, and rely on RVO (mandatory since C++17, but you'd be hard-pressed to find C++14 compiler that doesn't do it). Code as written is pretty confusing as to why you are doing this.

Reply:
Taking a first look at this it seems `CacheKey` can't have a constructor as `ParamsHash` requires it to be POD: https://github.com/pytorch/pytorch/blob/e6ef87001c2b7df1eac139adcaef174395e3a5a2/aten/src/ATen/native/utils/ParamsHash.h#L13

Reply:
Similar function for ConvolutionParams is called `setConvolutionParams`, with a comment on why it can't be a constructor, so it makes sense to follow this pattern here for consistency.